src/components/Pages/SetupPage/SetupPage.tsx

```diff
@@ -1,42 +1,118 @@
-  const onEmailChange = useCallback((_e: any, data: InputOnChangeData) => setEmail(data.value), [])
+  const redirectTo = useAfterLoginRedirection()
+
+  const onRandomize = useCallback(() => setProfile(getRandomDefaultProfile()), [])
+  const onContinue = useCallback(() => setView(View.FORM), [])
+  const onBack = useCallback(() => setView(View.RANDOMIZE), [])
+  const onNameChange = useCallback((_e: unknown, data: InputOnChangeData) => setName(data.value), [])
+  const onEmailChange = useCallback((_e: unknown, data: InputOnChangeData) => setEmail(data.value), [])
   const onAgreeChange = useCallback(() => setAgree(!agree), [agree])
 
-  const onSubmit = useCallback((e: React.FormEvent<HTMLFormElement>) => {
-    e.preventDefault()
-    setShowErrors(true)
+  const onSubmit = useCallback(
+    async (e: React.FormEvent<HTMLFormElement>) => {
+      e.preventDefault()
 
-    if (nameError || emailError || agreeError) {
-      return
-    }
+      setShowErrors(true)
 
-    // TODO: Handle submit.
-    console.log('submit')
-  }, [])
+      // If any of the fields has an error, don't submit.
+      if (nameError || emailError || agreeError) {
+        return
+      }
 
+      // These refs should have values at this point.
+      // If they don't, it means that there was something wrong on the initialization effect.
+      if (!accountRef.current || !identityRef.current) {
+        console.warn('No account or identity found.')
+        return
+      }
+
+      try {
+        setDeploying(true)
+
+        // Deploy a new profile for the user based on the selected default profile.
+        await deployProfileFromDefault({
+          connectedAccount: accountRef.current,
+          connectedAccountIdentity: identityRef.current,
+          defaultProfile: profile,
+          deploymentProfileName: name
+        })
+
+        // Subscribe to the newsletter only if the user has provided an email.
+        if (email) {
+          // Given that the subscription is an extra step, we don't want to block the user if it fails.
+          try {
+            await subscribeToNewsletter(email)
+          } catch (e) {
+            console.warn('There was an error subscribing to the newsletter', (e as Error).message)
+          }
+        }
+
+        // Redirect to the site defined in the search params.
+        if (redirectTo) {
+          window.location.href = decodeURIComponent(redirectTo)
+        } else {
+          window.location.href = '/'
+        }
+      } catch (e) {
+        setDeploying(false)
+
+        // TODO: Display a proper error on the page to show the user.
+        console.error('There was an error deploying the profile', (e as Error).message)
+      }
+    },
+    [nameError, emailError, agreeError, name, email, agree, profile, redirectTo]
+  )
+
+  // Validation effect.
+  // Will perform some validations on the form data whenever input changes.
   useEffect(() => {
+    // Validate the name.
```

Comment:
Instead of using a `useEffect` to check for all errors and set a value using a state hook, what do you think about 
using `useMemo` and storing the error in a variable? Doing so will perform the checks for each field whenever they change.
```
const nameError = useMemo(() => {
    if (!name.length) {
      return 'Please enter your name.'
    } else if (name.length >= 15) {
      return 'Name can have a maximum of 15 characters.'
    } else if (!/^[a-zA-Z0-9]+$/.test(name)) {
      return 'Name can only contain letters and numbers.'
    } else {
      return ''
    }
  
}, [name])
```

Reply:
Done